HW2/P2/parallel_vector.pyx

```diff
@@ -54,39 +54,58 @@ cpdef move_data_serial(np.int32_t[:] counts,
                        np.int32_t[:] src,
                        np.int32_t[:] dest,
                        int repeat):
-   cdef:
-       int idx, r
+    cdef:
+        int idx, r
 
-   assert src.size == dest.size, "Sizes of src and dest arrays must match"
-   with nogil:
-       for r in range(repeat):
-           for idx in range(src.shape[0]):
-               if counts[src[idx]] > 0:
-                   counts[dest[idx]] += 1
-                   counts[src[idx]] -= 1
+    assert src.size == dest.size, "Sizes of src and dest arrays must match"
+    with nogil:
+        for r in range(repeat):
+            for idx in range(src.shape[0]):
+                if counts[src[idx]] > 0:
+                    counts[dest[idx]] += 1
+                    counts[src[idx]] -= 1
 
 
 cpdef move_data_fine_grained(np.int32_t[:] counts,
                              np.int32_t[:] src,
                              np.int32_t[:] dest,
                              int repeat):
-   cdef:
-       int idx, r
-       omp_lock_t *locks = get_N_locks(counts.shape[0])
-
-   ##########
-   # Your code here
-   # Use parallel.prange() and a lock for each element of counts to parallelize
-   # data movement.  Be sure to avoid deadlock, and double-locking.
-   ##########
-   with nogil:
-       for r in range(repeat):
-           for idx in range(src.shape[0]):
-               if counts[src[idx]] > 0:
-                   counts[dest[idx]] += 1
-                   counts[src[idx]] -= 1
-
-   free_N_locks(counts.shape[0], locks)
+    cdef:
+        int idx, r,max_lock, min_lock 
+        omp_lock_t * locks = get_N_locks(counts.shape[0])
+        
+    ##########
+    # Your code here
+    # Use parallel.prange() and a lock for each element of counts to parallelize
+    # data movement.  Be sure to avoid deadlock, and double-locking.
+    ##########
+
+    #Causing deadlocks.
+    with nogil:
+        for r in range(repeat):
+            for idx in prange(src.shape[0],num_threads=4):  
+                if counts[src[idx]] > 0:
+                    '''
+                    #min max lock acquire release
+                    min_lock = min(dest[idx],src[idx])
+                    max_lock = max(dest[idx],src[idx])
+                    if min_lock != max_lock:
+                        acquire(&(locks[min_lock]))
+                        acquire(&(locks[max_lock]))
+                        counts[dest[idx]] += 1
+                        release(&(locks[dest[idx]]))
+                        counts[src[idx]] -= 1
+                        release(&(locks[src[idx]]))
+                    '''
+                    acquire(&(locks[dest[idx]]))
```

Comment:
Nice way to avoid deadlock. I did it by comparing dest[idx] and src[idx], but your way is more elegant